File: seven-governors-court-substrate-v0.1.0/scripts/build-package-manifest.mjs
import fs from "node:fs";
import path from "node:path";
import {
  PACKAGE_ROOT,
  canonicalCompact,
  canonicalJson,
  compareCodePoint,
  sha256,
  writeAtomic,
} from "./lib.mjs";

const mode = process.argv[2];
if (!new Set(["--check", "--emit"]).has(mode) || process.argv.length !== 3) {
  throw new Error("Specify exactly one of --check or --emit");
}
const excluded = new Set(["node_modules", "PACKAGE_MANIFEST.json"]);

function walk(directory, relative = "") {
  const records = [];
  for (const name of fs.readdirSync(directory).sort(compareCodePoint)) {
    if (excluded.has(name)) continue;
    const absolute = path.join(directory, name);
    const childRelative = path.posix.join(relative, name);
    const stat = fs.statSync(absolute);
    if (stat.isDirectory()) records.push(...walk(absolute, childRelative));
    else if (stat.isFile()) {
      const bytes = fs.readFileSync(absolute);
      records.push({ path: childRelative, bytes: bytes.length, sha256: sha256(bytes) });
    }
  }
  return records;
}

const files = walk(PACKAGE_ROOT);
const manifest = {
  schemaVersion: "1.0.0",
  packageName: "seven-governors-court-substrate",
  packageVersion: "0.1.0",
  releaseId: "court-substrate:0.1.0",
  releaseDate: "2026-08-09",
  manifestPolicy: "All package payload files except node_modules and this self-referential manifest.",
  fileCount: files.length,
  totalBytes: files.reduce((sum, item) => sum + item.bytes, 0),
  aggregateFingerprint: sha256(
    canonicalCompact(files.map((item) => [item.path, item.sha256])),
  ),
  files,
};
const text = canonicalJson(manifest);
const target = path.join(PACKAGE_ROOT, "PACKAGE_MANIFEST.json");
if (mode === "--emit") {
  writeAtomic(target, text);
  console.log(JSON.stringify({ status: "emitted", fileCount: files.length }));
} else if (fs.existsSync(target) && fs.readFileSync(target, "utf8") === text) {
  console.log(JSON.stringify({ status: "passed", fileCount: files.length }));
} else {
  console.error("STALE_PACKAGE_MANIFEST");
  process.exitCode = 1;
}

File: seven-governors-harmonic-invariants-v0.1.0/scripts/build_package_manifest.py
from __future__ import annotations

import argparse
import json
from pathlib import Path

from _bootstrap import PACKAGE_ROOT

from harmonic_invariants.canonical import (
    canonical_json_bytes,
    compact_json_bytes,
    sha256_bytes,
    sha256_file,
    write_atomic,
)


parser = argparse.ArgumentParser()
mode = parser.add_mutually_exclusive_group(required=True)
mode.add_argument("--check", action="store_true")
mode.add_argument("--emit", action="store_true")
arguments = parser.parse_args()
excluded = {"PACKAGE_MANIFEST.json", "__pycache__", ".pytest_cache"}
files = []
for path in sorted(PACKAGE_ROOT.rglob("*")):
    if not path.is_file() or any(part in excluded for part in path.parts):
        continue
    relative = path.relative_to(PACKAGE_ROOT).as_posix()
    files.append({"path": relative, "bytes": path.stat().st_size, "sha256": sha256_file(path)})
manifest = {
    "schemaVersion": "1.0.0",
    "packageName": "seven-governors-harmonic-invariants",
    "packageVersion": "0.1.0",
    "releaseId": "harmonic-invariants:0.1.0",
    "releaseDate": "2026-08-09",
    "manifestPolicy": "All package payload files except caches and this self-referential manifest.",
    "fileCount": len(files),
    "totalBytes": sum(item["bytes"] for item in files),
    "aggregateFingerprint": sha256_bytes(
        compact_json_bytes([[item["path"], item["sha256"]] for item in files])
    ),
    "files": files,
}
payload = canonical_json_bytes(manifest)
target = PACKAGE_ROOT / "PACKAGE_MANIFEST.json"
if arguments.emit:
    write_atomic(target, payload)
    print(json.dumps({"status": "emitted", "fileCount": len(files)}))
elif target.is_file() and target.read_bytes() == payload:
    print(json.dumps({"status": "passed", "fileCount": len(files)}))
else:
    print("STALE_PACKAGE_MANIFEST")
    raise SystemExit(1)
